Lib/ensurepip/__init__.py

```diff
@@ -1,26 +1,95 @@
-import os
+import collections
 import os.path
+import subprocess
 import sys
-import runpy
+import sysconfig
 import tempfile
-import subprocess
-from importlib import resources
-
-from . import _bundled
 
 
 
 __all__ = ["version", "bootstrap"]
+_PACKAGE_NAMES = ('setuptools', 'pip')
+_Package = collections.namedtuple('Package', ('version', 'filename'))
 
+# Directory of system wheel packages. Some Linux distribution packaging
+# policies recommand against bundling dependencies. For example, Fedora
+# installs wheel packages in the /usr/share/python-wheels/ directory and don't
+# install the ensurepip._bundled package.
+_WHEEL_PKG_DIR = sysconfig.get_config_var('WHEEL_PKG_DIR')
 
-_SETUPTOOLS_VERSION = "47.1.0"
 
-_PIP_VERSION = "20.2.3"
+def _get_versions(path):
+    versions = {}
+    try:
+        names = os.listdir(path)
+    except OSError:
+        # Ignore: path doesn't exist or permission error
+        names = ()
+    for name in names:
+        # name is like 'pip-20.2.3-py2.py3-none-any.whl'
+        if not name.endswith(".whl"):
+            continue
+        for package in _PACKAGE_NAMES:
+            prefix = package + '-'
+            if not name.startswith(prefix):
+                continue
+            part = name.removeprefix(prefix)
+            break
+        else:
+            continue
+
+        # Extract '20.2.3' from '20.2.3-py2.py3-none-any.whl'
+        version = []
+        part = part.split('-', 1)[0]
+        for part in part.split('.'):
+            try:
+                number = int(part)
+            except ValueError:
+                break
+            version.append(number)
+        if not version:
+            # failed to parse the version: ignore the package
+            continue
```

Comment:
This fails to parse rc/beta versions or post versions.

Reply:
https://www.python.org/dev/peps/pep-0632/ deprecates distutils.version, so I prefer to avoid it. I am not aware of any alternative in the stdlib: the PEP suggests to use the 3rd party packaging module.

I don't think that we use beta or rc version in ensurepip._bundled version. Is it the case in Fedora?

On Fedora, ensurepip._bundled is not installed, so it doesn't matter in practice :-)

Reply:
Oh, this only looks into ensurepip._bundled? OK